src/algebra/homology/homological_complex.lean

```diff
@@ -0,0 +1,657 @@
+  refl,
+end
+
+/-- If there is no `j` so `c.rel i j`, then `C.X_next i` is isomorphic to `0`. -/
+def X_next_iso_zero {i : ι} (h : c.next i = none) :
+  C.X_next i ≅ 0 :=
+eq_to_iso begin
+  dsimp [X_next],
+  rw h,
+  refl,
+end
+
+/--
+The differential mapping into `C.X j`, or zero if there isn't one.
+-/
+def d_to (j : ι) : C.X_prev j ⟶ C.X j :=
+match c.prev j with
+| none := (0 : C.X_prev j ⟶ C.X j)
+| (some ⟨i, w⟩) := (C.X_prev_iso w).hom ≫ C.d i j
+end
+
+/--
+The differential mapping out of `C.X i`, or zero if there isn't one.
+-/
+def d_from (i : ι) : C.X i ⟶ C.X_next i :=
+match c.next i with
+| none := (0 : C.X i ⟶ C.X_next i)
+| (some ⟨j, w⟩) := C.d i j ≫ (C.X_next_iso w).inv
+end
+
+lemma d_to_eq {i j : ι} (r : c.rel i j) :
+  C.d_to j = (C.X_prev_iso r).hom ≫ C.d i j :=
+begin
+  dsimp [d_to, X_prev_iso],
+  rw c.prev_eq_some r,
+  refl,
+end
+
+@[simp]
+lemma d_to_eq_zero {j : ι} (h : c.prev j = none) :
+  C.d_to j = 0 :=
+begin
+  dsimp [d_to],
+  rw h, refl,
+end
+
+lemma d_from_eq {i j : ι} (r : c.rel i j) :
+  C.d_from i = C.d i j ≫ (C.X_next_iso r).inv :=
+begin
+  dsimp [d_from, X_next_iso],
+  rw c.next_eq_some r,
+  refl,
+end
+
+@[simp]
+lemma d_from_eq_zero {i : ι} (h : c.next i = none) :
+  C.d_from i = 0 :=
+begin
+  dsimp [d_from],
+  rw h, refl,
+end
+
+@[simp, reassoc] lemma X_prev_iso_comp_d_to {i j : ι} (r : c.rel i j) :
+  (C.X_prev_iso r).inv ≫ C.d_to j = C.d i j :=
+by simp [C.d_to_eq r]
+
+@[simp, reassoc] lemma X_prev_iso_zero_comp_d_to {j : ι} (h : c.prev j = none) :
+  (C.X_prev_iso_zero h).inv ≫ C.d_to j = 0 :=
+by simp [h]
+
+@[simp, reassoc] lemma d_from_comp_X_next_iso {i j : ι} (r : c.rel i j) :
+  C.d_from i ≫ (C.X_next_iso r).hom = C.d i j :=
+by simp [C.d_from_eq r]
+
+@[simp, reassoc] lemma d_from_comp_X_next_iso_zero {i : ι} (h : c.next i = none) :
+  C.d_from i ≫ (C.X_next_iso_zero h).hom = 0 :=
+by simp [h]
+
+@[simp]
+lemma d_to_comp_d_from (j : ι) : C.d_to j ≫ C.d_from j = 0 :=
+begin
+  rcases h₁ : c.next j with _ | ⟨k,w₁⟩,
+  { rw [d_from_eq_zero _ h₁], simp, },
+  { rw [d_from_eq _ w₁],
+    rcases h₂ : c.prev j with _ | ⟨i,w₂⟩,
+    { rw [d_to_eq_zero _ h₂], simp, },
+    { rw [d_to_eq _ w₂], simp, } }
+end
+
+lemma kernel_from_eq_kernel [has_kernels V] {i j : ι} (r : c.rel i j) :
+  kernel_subobject (C.d_from i) = kernel_subobject (C.d i j) :=
+begin
+  rw C.d_from_eq r,
+  apply kernel_subobject_comp_mono,
+end
+
+lemma image_to_eq_image [has_images V] [has_equalizers V]
+  {i j : ι} (r : c.rel i j) :
+  image_subobject (C.d_to j) = image_subobject (C.d i j) :=
+begin
+  rw C.d_to_eq r,
+  apply image_subobject_iso_comp,
+end
+
+end
+
+namespace hom
+
+variables {C₁ C₂ C₃ : homological_complex V c}
+variables [has_zero_object V]
+local attribute [instance] has_zero_object.has_zero
+
+/-! Lemmas relating chain maps and `d_to`/`d_from`. -/
+
+/-- `f.f_prev j` is `f.f i` if there is some `r i j`, and zero otherwise. -/
+def f_prev (f : hom C₁ C₂) (j : ι) : C₁.X_prev j ⟶ C₂.X_prev j :=
```

Comment:
Would it make sense to use `prev` as the name? `f.prev` looks nicer than `f.f_prev`. Similarly for `f_next`.

Reply:
Great idea. Done